File: Midterm-Lab-3/MP20.js
const fs = require('fs');
const readline = require('readline');

/**
 * MP20 – Convert CSV dataset into JSON format.
 * Student: BOSITO, KALIEL OSH A.
 */

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

rl.question('Enter dataset file path: ', (filePath) => {
    if (!fs.existsSync(filePath)) {
        console.error('Error: File not found.');
        rl.close();
        return;
    }

    try {
        const fileContent = fs.readFileSync(filePath, 'utf8');
        const lines = fileContent.split(/\r?\n/);

        let headerFound = false;
        let headers = [];
        const jsonArray = [];

        for (let line of lines) {
            if (line.trim() === "" || (!headerFound && !line.includes("Candidate"))) {
                continue;
            }

            const cells = parseCsvLine(line);

            if (!headerFound) {
                headers = cells.map((h, i) => h.trim() || `column${i + 1}`);
                headerFound = true;
                continue;
            }

            const obj = {};
            headers.forEach((h, i) => {
                obj[h] = (cells[i] !== undefined ? cells[i].trim() : "") || "";
            });
            jsonArray.push(obj);
        }

        const json = JSON.stringify(jsonArray, null, 2);
        const preview = jsonArray.slice(0, 3);
        const previewJson = JSON.stringify(preview, null, 2);

        console.log("\n========= CSV TO JSON OUTPUT =========");
        console.log("Total records converted:", jsonArray.length);
        console.log("\n--- JSON (first 3 records shown) ---");
        console.log(previewJson);
        if (jsonArray.length > 3) {
            console.log(`... and ${jsonArray.length - 3} more records`);
        }
        console.log("=====================================");

    } catch (err) {
        console.error('Error processing file:', err.message);
    } finally {
        rl.close();
    }
});

function parseCsvLine(line) {
    const cells = [];
    let current = '';
    let inQuotes = false;

    for (let i = 0; i < line.length; i++) {
        const ch = line[i];

        if (ch === '"') {
            if (inQuotes && line[i + 1] === '"') {
                current += '"';
                i++;
            } else {
                inQuotes = !inQuotes;
            }
        } else if (ch === ',' && !inQuotes) {
            cells.push(current);
            current = '';
        } else {
            current += ch;
        }
    }

    cells.push(current);
    return cells;
}

File: Midterm-Lab-3/MP20.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Scanner;

/**
 * MP20 - Convert CSV dataset into JSON format.
 * Student: BOSITO, KALIEL OSH A.
 */
public class MP20 {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter dataset file path: ");
        String filePath = scanner.nextLine().trim();

        List<String> headers = new ArrayList<>();
        List<List<String>> rows = new ArrayList<>();

        try {
            loadDataset(filePath, headers, rows);
            runMP20(headers, rows);
        } catch (IOException e) {
            System.out.println("Error: Unable to read file. " + e.getMessage());
        } catch (IllegalArgumentException e) {
            System.out.println("Error: " + e.getMessage());
        } finally {
            scanner.close();
        }
    }

    private static void runMP20(List<String> headers, List<List<String>> rows) {
        List<Map<String, String>> jsonArray = new ArrayList<>();

        for (List<String> row : rows) {
            Map<String, String> obj = new LinkedHashMap<>();
            for (int i = 0; i < headers.size(); i++) {
                String key = headers.get(i).trim();
                if (key.isEmpty()) {
                    key = "column" + (i + 1);
                }
                String value = (i < row.size()) ? row.get(i).trim() : "";
                obj.put(key, value);
            }
            jsonArray.add(obj);
        }

        String json = toJsonString(jsonArray);
        System.out.println("\n========= CSV TO JSON OUTPUT =========");
        System.out.println("Total records converted: " + jsonArray.size());
        System.out.println("\n--- JSON (first 3 records shown) ---");
        System.out.println(json);
        System.out.println("=====================================");
    }

    private static String toJsonString(List<Map<String, String>> jsonArray) {
        StringBuilder sb = new StringBuilder("[\n");
        int limit = Math.min(3, jsonArray.size());
        for (int i = 0; i < limit; i++) {
            Map<String, String> obj = jsonArray.get(i);
            sb.append("  {");
            int j = 0;
            for (Map.Entry<String, String> entry : obj.entrySet()) {
                if (j > 0) sb.append(", ");
                sb.append("\"").append(escapeJson(entry.getKey())).append("\": ");
                sb.append("\"").append(escapeJson(entry.getValue())).append("\"");
                j++;
            }
            sb.append("}");
            if (i < limit - 1) sb.append(",");
            sb.append("\n");
        }
        if (jsonArray.size() > 3) {
            sb.append("  ... and ").append(jsonArray.size() - 3).append(" more records\n");
        }
        sb.append("]");
        return sb.toString();
    }

    private static String escapeJson(String s) {
        if (s == null) return "";
        return s.replace("\\", "\\\\")
                .replace("\"", "\\\"")
                .replace("\n", "\\n")
                .replace("\r", "\\r")
                .replace("\t", "\\t");
    }

    private static void loadDataset(String filePath, List<String> headers, List<List<String>> rows)
            throws IOException {
        try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
            String line;
            boolean headerFound = false;

            while ((line = reader.readLine()) != null) {
                if (line.trim().isEmpty()) {
                    continue;
                }

                List<String> cells = parseCsvLine(line);

                if (!headerFound) {
                    boolean isHeader = false;
                    for (String cell : cells) {
                        if ("candidate".equalsIgnoreCase(cell.trim())) {
                            isHeader = true;
                            break;
                        }
                    }

                    if (isHeader) {
                        headers.addAll(cells);
                        headerFound = true;
                    }
                    continue;
                }

                rows.add(cells);
            }
        }

        if (headers.isEmpty()) {
            throw new IllegalArgumentException("Dataset header row not found.");
        }
    }

    private static List<String> parseCsvLine(String line) {
        List<String> cells = new ArrayList<>();
        StringBuilder current = new StringBuilder();
        boolean inQuotes = false;

        for (int i = 0; i < line.length(); i++) {
            char ch = line.charAt(i);

            if (ch == '"') {
                if (inQuotes && i + 1 < line.length() && line.charAt(i + 1) == '"') {
                    current.append('"');
                    i++;
                } else {
                    inQuotes = !inQuotes;
                }
            } else if (ch == ',' && !inQuotes) {
                cells.add(current.toString());
                current.setLength(0);
            } else {
                current.append(ch);
            }
        }

        cells.add(current.toString());
        return cells;
    }
}